source/slang/ir.cpp

```diff
@@ -4536,6 +4562,308 @@ namespace Slang
+            // to represent the parameter, and then perform a load
+            // form it at the start of the function.
+            //
+            auto valueType = paramType;
+            auto globalVariable = addGlobalVariable(builder->getModule(), valueType);
+            builder->addLayoutDecoration(globalVariable, paramLayout);
+            moveValueBefore(globalVariable, builder->getFunc());
+
+            auto irLoad = builder->emitLoad(globalVariable);
+            pp->replaceUsesWith(irLoad);
+        }
+
+    }
+
+    void legalizeEntryPointParameterForGLSL(
+        GLSLLegalizationContext*    context,
+        IRFunc*                     func,
+        IRParam*                    pp,
+        VarLayout*                  paramLayout)
+    {
+        auto builder = context->getBuilder();
+        auto stage = context->getStage();
+
+        // We need to create a global variable that will replace the parameter.
+        // It seems superficially obvious that the variable should have
+        // the same type as the parameter.
+        // However, if the parameter was a pointer, in order to
+        // support `out` or `in out` parameter passing, we need
+        // to be sure to allocate a variable of the pointed-to
+        // type instead.
+        //
+        // We also need to replace uses of the parameter with
+        // uses of the variable, and the exact logic there
+        // will differ a bit between the pointer and non-pointer
+        // cases.
+        auto paramType = pp->getDataType();
+
+        // First we will special-case stage input/outputs that
+        // don't fit into the standard varying model.
+        // For right now we are only doing special-case handling
+        // of geometry shader output streams.
+        if( auto paramPtrType = as<IROutTypeBase>(paramType) )
+        {
+            auto valueType = paramPtrType->getValueType();
+            if( auto gsStreamType = as<IRHLSLStreamOutputType>(valueType) )
+            {
+                // An output stream type like `TriangleStream<Foo>` should
+                // more or less translate into `out Foo` (plus scalarization).
+
+                auto globalOutputVal = createGLSLGlobalVaryings(
+                    context,
+                    builder,
+                    valueType,
+                    paramLayout,
+                    LayoutResourceKind::VaryingOutput,
+                    stage);
+
+                // TODO: a GS output stream might be passed into other
+                // functions, so that we should really be modifying
+                // any function that has one of these in its parameter
+                // list (and in the limit we should be leagalizing any
+                // type that nests these...).
+                //
+                // For now we will just try to deal with `Append` calls
+                // directly in this function.
+
+
+
+                for( auto bb = func->getFirstBlock(); bb; bb = bb->getNextBlock() )
```

Comment:
Why does this do the search looking for the GLSL intrnisic for EmitVertex()? Because we can't search for append on the right type? 

Also it appears as if it's searching for such a call by searching blocks, and then instructions in blocks. But can't I have nested blocks? Is this because its SSA at this point and flattened..?

Reply:
A few points:
- This is all the original code that got hoisted out to a function rather than new code. It is fine to review it, but I just want to note that it wasn't added just now.
- The reason to look for the intrinsic decoration is that it will have an easy-to-match string, while looking for the function name `Append` is actually hard, because that would be mixed into some big mangled name
- Also, this translation for geometry shaders isn't really complete since it doesn't account for `Append` calls from subroutines, or multiple GS output streams. It needs work.

Reply:
Oh, and on the actual semantic question: our IR doesn't have nested blocks. A function has blocks as children, and blocks have instructions as children, and that is true throughout the lifetime of the IR (except when you have a function nested in a generic, but then both the generic and the function are "flat" on their own). This search should work for `Append` calls in any entry point function, although as noted above it doesn't account for calls from other functions that we call.